### ISSUE TITLE
Make register_on_player_receive_fields also receive the fields that did not change

### ISSUE THREAD

**Issue by Bastrabun** [minetest/minetest#10223]
##### Minetest version
```
5.3.0
```

##### OS / Hardware
Operating system: Debian 10
CPU: Intel(R) Core(TM) i7 CPU         950  @ 3.07GHz

GPU model: Radeon HD 7870 GHz Edition
OpenGL version: 4.5

##### Summary

Some formspec elements do not get sent to register_on_player_receive_fields when a button is clicked, some others do. textareas seem to be sent always, checkboxes and dropdowns only when changed.

##### Steps to reproduce
1. display this formspec to a player:

```
formspec_version[3]
size[5,5]
checkbox[0.2,0.2;my_checkbox;Checkbox!;true]
textarea[0.2,0.8;2,2;my_textarea;Textarea!;Some text]
button[0.2,4;3,0.75;button;Go]
```

2. catch it with register_on_player_receive_fields

3. dump(fields)

=> Only when I click the checkbox this shows up in the fields. When I hit the button, the checkbox does not get sent together with the other fields

Expectation is that ALL form members get sent and can be accessed in the register_on_player_receive_fields

Here's a testmod, please take a look at its log.

```
local fs = "formspec_version[3]size[5,5]checkbox[0.2,0.2;my_checkbox;Checkbox!;true]textarea[0.2,0.8;2,2;my_textarea;Textarea!;Some text]button[0.2,4;3,0.75;button;Go]"

minetest.register_chatcommand("show_me", {
    func = function( name , param)
        minetest.show_formspec(name, "mymod:myformspec", fs)
    end,
})

minetest.register_on_player_receive_fields(function(player, formname, fields)

    if formname ~= "mymod:myformspec" then
        return
    end

    minetest.log("action",dump(fields))
end)
```

Here's the log if I toggle the checkbox:

```
2020-07-24 04:53:47: ACTION[Server]: {
	my_textarea = "Some textasdasdsad",
	my_checkbox = "false"
}
```

Here's the log when I hit the go-button:

```
2020-07-24 04:53:48: ACTION[Server]: {
	my_textarea = "Some textasdasdsad",
	button = "Go"
}
```

**Comment by v-rob** [comment]
It's not a bug since it's well-defined behaviour in `lua_api.txt`, but I do agree that it would be nice to have it received for everything.  If this is implemented, callbacks should (and would have to) be refined in general for better usage and more consistency.

**Comment by Wuzzy2** [comment]
As a modder and gamedev, IMO the most annoying thing about `player_receive_fields` is that checkboxes are only sent when they are toggled. If they were not toggled, you don't get to know the checkbox state.

This forces you to remember the checkbox state which blows up your code a bit. Sure, it's easy enough to do, but it's dumb boilerplate that could be easily avoided. Less code is better. :P

**Comment by SmallJoker** [comment]
https://github.com/minetest/minetest/blob/808fa5ecb3ddfd5d993000cc6b4c972257e182db/src/gui/guiFormSpecMenu.cpp#L3745-L3747

This check could be removed, but might cause unintended side-effects. Sometimes you want to capture changes only - for example scrollbars that cause a formspec update. The formspec will update with each dropdown input and needs manual change-checking.

**Comment by sorcerykid** [comment]
Any change to this behavior has the potential to break a lot of mods as it requires completely different logic for form processing, so something like this will need to be given a great deal of forethought.

**Comment by hecktest** [comment]
Having to cache values is only annoying until you realize that you should be wrapping formspecs in your code anyway, not writing and handling them raw which is very brittle and error prone.

Sending all fields will break all existing GUI toolkits that rely on the already documented behavior, and force them to waste cycles comparing all the values instead to see what changed.

**Comment by v-rob** [comment]
> 
> 
> Any change to this behavior has the potential to break a lot of mods as it requires completely different logic for form processing, so something like this will need to be given a great deal of forethought.

Precisely.  At the very least, changing the behaviour needs an extra element, e.g. `new_events[]` to turn it on.  Personally, I would prefer to wait until when we're getting a new framework, i.e. whatever #6527 finally decides on, before events are changed.  Currently, things are doable, and that's good enough for now.

**Comment by hecktest** [comment]
Even with a new GUI framework, it would make zero sense to have the server receive and filter values that it already knows. It's bad design.

**Comment by v-rob** [comment]
> The current approach makes the most sense.

Not really, because there _is_ no current method.  We have multiple methods of sending for multiple reasons, and none are very consistent with each other:
* Buttons, checkboxes, hypertext, tabheaders, tables, and textlists send their own event
* Animated images, (pwd)fields, and textareas are sent when a different event is triggered
* Scrollbars and dropdowns are a combination: If the dropdown/scrollbar gives an event, only that single dropdown/scrollbar gives the event.  If another event happens, e.g. a button press, _all_ scrollbars and dropdowns give their status.

I think the main problem with formspec events lies here:

> Having to cache values is only annoying until you realize that you should be wrapping formspecs in your code anyway, not writing and handling them raw which is very brittle and error prone.

This is very true; you do have to cache values, and this is **not** something that mods should have to do themselves.  This is why I pointed to #6527, because the formspec _format_ should cache the values itself without the modder having to do it manually.  I don't know how, but it should.

**Comment by sorcerykid** [comment]
> This is very true; you do have to cache values, and this is not something that mods should have to do themselves. This is why I pointed to #6527, because the formspec format should cache the values itself without the modder having to do it manually. I don't know how, but it should.

One of my long term goals with ActiveFormspecs is to track the states of all elements, thereby alleviating the inconsistency with event handling. I might get around to finally implementing that in the next version since there seems to be interest.

**Comment by hecktest** [comment]
> I don't know how, but it should.

We could add some unary operator like #, @, $ or ! that marks a form element as always sent. This would help OP in whatever he's trying to do, without ruining the efficiency of the current system. You could even emulate the send with a cached value so that no bandwidth is wasted on this. (Just in case - do not use that as an argument that we could just emulate everything. The time wasted iterating over unwanted values would still be tragic. Only the game knows what it actually needs!)

**Comment by Zughy** [comment]
Another problem with this is that there currently is no cohesion between elements: in a format like the one below...
![imagen](https://user-images.githubusercontent.com/63455151/147391595-be137f02-7d9d-46bd-b170-80d099b7e321.png)
clicking a checkbox sends the state of the checkbox AND every dropdown field. But changing a dropdown entry only sends the current dropdown field. Why? I understand that clicking a button sends everything as they're usually need to confirm something, but checkboxes?

**Comment by doxygen-spammer** [comment]
If you read user interface design manuals, you will quickly learn that checkboxes are not meant for immediate actions, but for configuration.
It may be possible that a Minetest mod wants to store the configuration somewhere else, immediately after the checkbox was toggled.
But it is equally valid if the checkbox does not have any meaning outside of this specific form, and only gets a meaning when another button is pressed.

Take a mailbox or shopping basket as example:
You can check several items.
But the checked state has only a meaning when a button like “mark as read” or “remove” is pressed.
After closing the form, the checked states will never have a meaning again.
In this case, it is just useless to disconnect checkboxes from buttons.

---

The manual does not make any distinction between comboboxes, text fields, and checkboxes.
All it says is this:

> * Fields are sent for formspec elements which define a field. `fields`
> is a table containing each formspecs element value (as string), with
> the `name` parameter as index for each. The value depends on the
> formspec element type:

But it does make a distinction between buttons and checkboxes:

> * `button` and variants: If pressed, contains the user-facing button
> text as value. **If not pressed, is `nil`**

The element `checkbox` does not have the “if not pressed” exception, it just says:

> * `checkbox`: `"true"` if checked, `"false"` if unchecked

(Note that “checked” is not only a verb, but also an adjective.)

---

> Any change to this behavior has the potential to break a lot of mods as it requires completely different logic for form processing, so something like this will need to be given a great deal of forethought.

This makes no sense to me.

1. Formspec already has the `formspec_version[]` element. I do not see a reason to modify the behavior with the introduction of a new version.
2. Mods are expected to verify form submissions. This is even written in the manual. If a user would open the form, press the checkbox twice, and only the second press is sent to the mod (for whatever reason), the mod would already be broken if it could not handle that. I do not see the reason to support broken mods, but not mods which do it right.
3. What prevents a mod to capture any new checkbox value, and then remember this value until something else than a checkbox is pressed?

---

Another thing:

> It's not a bug since it's well-defined behaviour in lua_api.txt,

Please reevaluate your comprehension of things like a reference manual.
The purpose of a reference manual is not to write it ambiguously, so you can bend the meaning whenever a bug report upsets you.
I observe this misuse of reference documentation very rarely, but far too often in Minetest. :(

A reference manual, like lua_api.txt, is often the only source of information, on which a user of the software can rely.
If the manual carries wrong information, it is exactly this: a bug in the manual.
If the manual carries correct information, but the software behaves differently, it is a bug in the software.

In this case, the manual is missing important information, which would not be important if the software would behave in the way suggested by the available information.
You can now choose if it is a bug in the manual or in the software, but *it is a bug*.

**Comment by v-rob** [comment]
> Please reevaluate your comprehension of things like a reference manual.

If people can't rely on the reference manual, the only thing they can rely on is empirical trial and error.  That's many, many times worse, since then they rely on bugs; there are way too many examples of this in Minetest, and the results of it are the subject of great debate.  We sometimes have errors where lua_api.txt is wrong; this is not one of them.  This has been the expected behaviour for years.  Admittedly, the documentation is spotty at best, so modders have had to rely on empiricism to get the full story, but that doesn't make this a bug.

> If you read user interface design manuals, you will quickly learn that checkboxes are not meant for immediate actions, but for configuration.

That doesn't mean they shouldn't send a field.  When you click on a checkbox, it might do other things, such as disable other checkboxes when they are incompatible with it.  For instance, try disabling the "Shaders" checkbox in the main menu settings--it disables the fine shader checkboxes.  Hence, they deserve to have a field sent on click.

> Mods are expected to verify form submissions.

This is a disconnect from reality.  Regardless of what they're supposed to do, they do not always do so.  In any case, there's no requirement anywhere that they don't rely on precisely which fields are sent by clicking which elements, so changing the behaviour would be an error on our part, not theirs.  Their way might be brittle and inadvisable, but not wrong.

---

In any case, it's quite clear that the only way for this to happen is an opt-in element, just like formspecs have been doing for some time now for new features.

**Comment by rubenwardy** [comment]
I agree that this isn't a bug, as it's how formspecs have been designed. It is inconsistent thought.

This isn't necessarily a breaking change as you can put it behind a formspec version or config element, or you could put all the values into a separate table passed to callbacks.

You need a way to indicate that submission is due to a checkbox change though. I think a better API for a future formspec version would be to submit all fields, and then include fields about the event (`event_from` (element name), `event_type` (value_changed, etc))

**Comment by jordan4ibanez** [comment]
Also it should probably send you the data when the formspec quits
